pkg/cmd/image/list.go

```diff
@@ -188,50 +189,126 @@ type imagePrinter struct {
 	client                                 *containerd.Client
-	contentStore                           content.Store
+	contentStore                           containerdutil.Provider
 	snapshotter                            snapshots.Snapshotter
 }
 
-func (x *imagePrinter) printImage(ctx context.Context, img images.Image) error {
-	ociPlatforms, err := images.Platforms(ctx, x.contentStore, img.Target)
+type Image struct {
+	blobSize int64
+	size     int64
+	platform platforms.Platform
+	config   *ocispec.Descriptor
+}
+
+var ReadBlob = containerdutil.WithCache()
+
+func ReadManifest(ctx context.Context, provider containerdutil.Provider, snapshotter snapshots.Snapshotter, desc ocispec.Descriptor) (*Image, error) {
+	// Read the manifest blob from the descriptor
+	manifestData, err := ReadBlob(ctx, provider, desc)
 	if err != nil {
-		log.G(ctx).WithError(err).Warnf("failed to get the platform list of image %q", img.Name)
-		return x.printImageSinglePlatform(ctx, img, platforms.DefaultSpec())
+		return nil, err
 	}
-	psm := map[string]struct{}{}
-	for _, ociPlatform := range ociPlatforms {
-		platformKey := makePlatformKey(ociPlatform)
-		if _, done := psm[platformKey]; done {
-			continue
-		}
-		psm[platformKey] = struct{}{}
-		if err := x.printImageSinglePlatform(ctx, img, ociPlatform); err != nil {
-			log.G(ctx).WithError(err).Warnf("failed to get platform %q of image %q", platforms.Format(ociPlatform), img.Name)
-		}
+
+	// Unmarshal as Manifest
+	var manifest ocispec.Manifest
+	if err := json.Unmarshal(manifestData, &manifest); err != nil {
+		return nil, err
 	}
-	return nil
+
+	// Now, read the config
+	configData, err := ReadBlob(ctx, provider, manifest.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal as Image
+	var config ocispec.Image
+	if err := json.Unmarshal(configData, &config); err != nil {
+		log.G(ctx).Error("Error unmarshaling config")
+		return nil, err
+	}
+
+	// If we are here, the image exists and is valid, so, do our size lookups
+
+	// Aggregate the blob size from the config and layers
+	blobSize := manifest.Config.Size
+	for _, layerDescriptor := range manifest.Layers {
+		blobSize += layerDescriptor.Size
+	}
+
+	// Get the platform
+	plt := platforms.Normalize(ocispec.Platform{OS: config.OS, Architecture: config.Architecture, Variant: config.Variant})
+
+	// Get the filesystem size for all layers
+	chainID := identity.ChainID(config.RootFS.DiffIDs).String()
+	size := int64(0)
+	if _, actualSize, err := imgutil.ResourceUsage(ctx, snapshotter, chainID); err == nil {
+		size = actualSize.Size
+	}
+
+	return &Image{
+		blobSize: blobSize,
+		size:     size,
+		platform: plt,
+		config:   &manifest.Config,
+	}, nil
 }
 
-func makePlatformKey(platform v1.Platform) string {
-	if platform.OS == "" {
-		return "unknown"
+func ReadIndex(ctx context.Context, provider containerdutil.Provider, snapshotter snapshots.Snapshotter, desc ocispec.Descriptor) (map[string]*Image, error) {
```

Comment:
Does this need to be exported?

Reply:
I believe as a follow-up PR we should move these methods to some `util` package and make use of them instead of what we are currently doing.
This will definitely benefit other commands (like `rmi`) which lookup methods are currently faulty (see for example #3016 (that should stay open btw)).

If you prefer me to un-export these for now, until we move these for more general use, I am of course happy to do that.

Reply:
Un-exported with next push.
We can always make them public when need be later on.